loadbalancer/loadbalancer.go

```diff
@@ -0,0 +1,367 @@
+		if request.Kind == messages.RoutingKindTerminate {
+			expiryPtr = nil
+		}
+		err = messages.WriteRoutingResponseSuccess(s, expiryPtr)
+	}
+
+	// log any errors writing the response
+	if err != nil {
+		log.Warnf("writing routing response: %s", err)
+		return
+	}
+}
+
+func (lb *LoadBalancer) processRoutingRequest(p peer.ID, request *messages.RoutingRequest, expiry time.Time) error {
+	lb.routesLk.Lock()
+	defer lb.routesLk.Unlock()
+	switch request.Kind {
+	case messages.RoutingKindNew:
+		// handle new requests
+		return lb.setupNewRoutes(p, request.Protocols, expiry)
+	case messages.RoutingKindExtend:
+		// handle route extensions
+		return lb.extendRoutes(p, request.Protocols, expiry)
+	case messages.RoutingKindTerminate:
+		// handle terminates
+		return lb.terminateRoutes(p, request.Protocols)
+	default:
+		// this case really shouldn't happen -- it should fail deserializing the request
+		return errors.New("unrecognized request kind")
+	}
+}
+
+func (lb *LoadBalancer) setupNewRoutes(p peer.ID, protocols []protocol.ID, expiry time.Time) error {
+
+	// check route filter if present to verify if we can setup this route
+	if lb.newRouteFilter != nil {
+		err := lb.newRouteFilter(p, protocols)
+		if err != nil {
+			return err
+		}
+	}
+
+	// check existing routes
+	for _, id := range protocols {
+		if _, ok := lb.routes[id]; ok {
+			// error if this protocol is already registered
+			return ErrAlreadyRegistered{id}
+		}
+	}
+
+	// setup routes
+	for _, id := range protocols {
+		id := id
+		expiryTimer := lb.clock.AfterFunc(expiry.Sub(lb.clock.Now()), func() {
+			lb.routesLk.Lock()
+			lb.expireRoute(id, false)
+			lb.routesLk.Unlock()
+		})
+		lb.peers[p] = append(lb.peers[p], &registeredRoute{id, expiryTimer})
+		lb.routes[id] = p
+		lb.h.SetStreamHandler(id, lb.handleIncoming)
+	}
+	return nil
+}
+
+func (lb *LoadBalancer) extendRoutes(p peer.ID, protocols []protocol.ID, expiry time.Time) error {
+
+	// check routes to verify ownership
+	for _, id := range protocols {
+		registeredPeer, ok := lb.routes[id]
+		if !ok || p != registeredPeer {
+			// error if this protocol not registered to this peer
+			return ErrNotRegistered{p, id}
+		}
+	}
+
+	// setup routes
+	for _, id := range protocols {
+		id := id
+		for _, route := range lb.peers[p] {
+			if route.protocol == id {
+				// TODO: is there a race around the timer stopping but the after func not yet being called?
+				route.expiryTimer.Stop()
+				route.expiryTimer = lb.clock.AfterFunc(expiry.Sub(lb.clock.Now()), func() {
+					lb.routesLk.Lock()
+					lb.expireRoute(id, false)
+					lb.routesLk.Unlock()
+				})
+				break
+			}
+		}
+	}
+	return nil
+}
+
+func (lb *LoadBalancer) terminateRoutes(p peer.ID, protocols []protocol.ID) error {
+
+	// check routes to verify ownership
+	for _, id := range protocols {
+		registeredPeer, ok := lb.routes[id]
+		if !ok || p != registeredPeer {
+			// error if this protocol not registered to this peer
+			return ErrNotRegistered{p, id}
```

Comment:
is this a bit too harsh for a "remove them all please!"? should we instead terminate any that exist but also return an `ErrNotRegistered` if any of those didn't exist rather than bail if just one doesn't exist?